File: core/CoreDatabaseManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.core;

import grakn.client.api.database.Database;
import grakn.client.api.database.DatabaseManager;
import grakn.client.common.GraknClientException;
import grakn.client.common.GraknStub;

import java.util.List;

import static grakn.client.common.ErrorMessage.Client.DB_DOES_NOT_EXIST;
import static grakn.client.common.ErrorMessage.Client.MISSING_DB_NAME;
import static grakn.client.common.RequestBuilder.Core.DatabaseManager.allReq;
import static grakn.client.common.RequestBuilder.Core.DatabaseManager.containsReq;
import static grakn.client.common.RequestBuilder.Core.DatabaseManager.createReq;
import static java.util.stream.Collectors.toList;

public class CoreDatabaseManager implements DatabaseManager {

    private final CoreClient client;

    public CoreDatabaseManager(CoreClient client) {
        this.client = client;
    }

    @Override
    public Database get(String name) {
        if (contains(name)) return new CoreDatabase(this, name);
        else throw new GraknClientException(DB_DOES_NOT_EXIST, name);
    }

    @Override
    public boolean contains(String name) {
        return stub().databasesContains(containsReq(nonNull(name))).getContains();
    }

    @Override
    public void create(String name) {
        stub().databasesCreate(createReq(nonNull(name)));
    }

    @Override
    public List<CoreDatabase> all() {
        List<String> databases = stub().databasesAll(allReq()).getNamesList();
        return databases.stream().map(name -> new CoreDatabase(this, name)).collect(toList());
    }

    GraknStub.Core stub() {
        return client.stub();
    }

    static String nonNull(String name) {
        if (name == null) throw new GraknClientException(MISSING_DB_NAME);
        return name;
    }
}

File: core/CoreDatabaseManager.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {DatabaseManager} from "../api/database/DatabaseManager";
import {Database} from "../api/database/Database";
import {TypeDBClientError} from "../common/errors/TypeDBClientError";
import {ErrorMessage} from "../common/errors/ErrorMessage";
import {RequestBuilder} from "../common/rpc/RequestBuilder";
import {CoreDatabase} from "./CoreDatabase";
import {TypeDBClient} from "typedb-protocol/core/core_service_grpc_pb";

export class CoreDatabaseManager implements DatabaseManager {

    private readonly _rpcClient: TypeDBClient;

    constructor(client: TypeDBClient) {
        this._rpcClient = client;
    }

    public async get(name: string): Promise<Database> {
        if (await this.contains(name)) return new CoreDatabase(name, this._rpcClient);
        else throw new TypeDBClientError(ErrorMessage.Client.DB_DOES_NOT_EXIST);
    }

    public create(name: string): Promise<void> {
        if (!name) throw new TypeDBClientError(ErrorMessage.Client.MISSING_DB_NAME);
        const req = RequestBuilder.Core.DatabaseManager.createReq(name);
        return new Promise((resolve, reject) => {
            this._rpcClient.databases_create(req, (err) => {
                if (err) reject(new TypeDBClientError(err));
                else resolve();
            });
        });
    }

    public contains(name: string): Promise<boolean> {
        if (!name) throw new TypeDBClientError(ErrorMessage.Client.MISSING_DB_NAME);
        const req = RequestBuilder.Core.DatabaseManager.containsReq(name);
        return new Promise((resolve, reject) => {
            this._rpcClient.databases_contains(req, (err, res) => {
                if (err) reject(new TypeDBClientError(err));
                else resolve(res.getContains());
            });
        });
    }

    public all(): Promise<Database[]> {
        const req = RequestBuilder.Core.DatabaseManager.allReq();
        return new Promise((resolve, reject) => {
            this._rpcClient.databases_all(req, (err, res) => {
                if (err) reject(new TypeDBClientError(err));
                else resolve(res.getNamesList().map(name => new CoreDatabase(name, this._rpcClient)));
            })
        })
    }

    rpcClient() {
        return this._rpcClient;
    }

}